### ISSUE TITLE
iOS Swift Package Manager (SPM) support

### ISSUE THREAD

**Issue by zoltancsipler** [google-ai-edge/mediapipe#5464]
### OS Platform and Distribution

iOS

### Programming Language and version

Swift

### Android AAR (if android)

None

### Describe the problem

Cannot integrate mediapipe libraries with SPM

### Complete Logs

```shell
On iOS, are you planning to add SPM support to mediapipe libs? Is there any way currently to integrate mediapipe libraries for apps that are using spm and not using cocoapods?
```

**Comment by zoltancsipler** [comment]
On iOS, are you planning to add SPM support to mediapipe libs? Is there any way currently to integrate mediapipe libraries for apps that are using spm and not using cocoapods?

**Comment by mantoljak** [comment]
Yes I am also interested in SPM support as i use local modules as SwiftPackage, not cocoapods. The problem is when i add MediaPipe pod i cannot import it into my spm unless i create a bridging framework which is basically duplicating everything.

**Comment by jurajantas** [comment]
Would be cool to have manual install or SPM. Cocoa pods are causing issues with SwiftUI previews. (preview crashes)

**Comment by jurajantas** [comment]
Guys, are there plans to move to SPM or not? This is critical for us to make decision as CocoaPods used for distribution is liability for us. SPM or raw .xcframework would be fine.

**Comment by mantoljak** [comment]
That is why this company (Google) is bound to fail. Nobody in the company with 100k fucking employees can answer to this simple question. They probably need to have 33 meetings to discuss what they will answer here. It is pathetic.

**Comment by echamussy** [comment]
Any news on this? Cocoapods is basically deprecated at this point and it does not work well with XCode 16 and SwiftUI. Please provide a way to integrate via SPM or manually. Thanks.

**Comment by dohque** [comment]
Voting up for this one. We may have to seek alternatives for MediaPipe if there will be no SPM support.

**Comment by paescebu** [comment]
I don't know if it suits your needs. And probably Google will end up bringing their own Swift package. Especially now that CocoaPods is deprecated.
But i tried to package it all up into a Swift package. A GitHub Action updates the frameworks automatically every Sunday if there are changes on Googles side.

My goal was to not introduce any wrapper types or anything. Was tricky and took me some time. But it works for me now.

https://github.com/paescebu/SwiftTasksVision

And who knows, now that its out, maybe it forces Google to sherlock this and provide their own "proper" Swift package.

**Comment by jurajantas** [comment]
According to linkedin assignee Maxime Brenon no longer works for Google.

**Comment by riderx** [comment]
@paescebu any plan to release for chat as well ?

**Comment by paescebu** [comment]
@riderx Heyy,
I was always open to spin up the same solution for other MediaPipe Tasks. 
But first wanted to see what the traction/interest was like for the first one.
~~What are you referring to when you say "Chat"~~? :)

UPDATE:
Ah got it, you probably meant LLM inference task!
Thats definitely a very interesting one, the interest for it could be very high. Let me see if I can do the same in the close future. Will surely keep you up to date if i manage. I think something worth to try for me next week :)

**Comment by paescebu** [comment]
@riderx there you go:
https://github.com/paescebu/SwiftTasksGenAI

Unfortunately its not as convenient as the Vision one, as some binaries were larger than 100MB. This forced me to use Git LFS, which means you cant use it as a remote package in Swift package manager (SPM limitations). BUT you can still use it as a local Swift package the same way.

As soon as Apple supports Git LFS properly with SPM this package will just work without changes (apart from README updates).

let me know i it works for you, if you have issues with this quick solution, feel free to file an issue on the project.

Regards and a great Sunday
Pascal

**Comment by paescebu** [comment]
@riderx 
The local package solution sucked.

Now i switched to Release artifacts and remote binary targets that SPM pulls, so no more reliance on Git-LFS. 
Should have done that from the beginning.

Now you can use it just as you could with `SwiftTasksVision` and have it setup as a remote dependency within Xcode or another Swift Package.
The ExampleApp shows how it works for Xcode.

Regards and a nice weekend!

**Comment by paescebu** [comment]
@riderx 
And now it supports also semantic versioning. No more limitations at all. Gonna apply it to the Vision Package now as well.
Who would have known making an additional target depending on the target with unsafe flags is suddenly fine for Xcode :D. TIL.

UPDATE: Also SwiftTasksVision now supports semantic versioning

...I guess now I also have to do MediaPipe Task Text 😆... Should be  straight forward now

**Comment by paescebu** [comment]
I guess now I also have to do Text 😆... Should be equally straight forward now

**Comment by paescebu** [comment]
And now I added also `SwiftTasksText` for MediaPipeTasksText.
Now all iOS Compatible MediaPipeTasks should be supported and easy to integrate:
https://github.com/paescebu/SwiftTasksText

**Comment by paescebu** [comment]
And last but not least, a fork of the samples repo with all iOS Samples updated by using those Swift Packages above instead: 
https://github.com/paescebu/mediapipe-samples-spm

All manually tested and seem to work just as expected :)

**Comment by riderx** [comment]
Man you are a machine ! sorry for been silent I'm adding it now to https://github.com/Cap-go/capacitor-llm

**Comment by riderx** [comment]
@paescebu i'm not sure i got it properly does my doc in https://github.com/Cap-go/capacitor-llm is correct or i could directly link the lib in my package ?

**Comment by paescebu** [comment]
> @paescebu i'm not sure i got it properly does my doc in https://github.com/Cap-go/capacitor-llm is correct or i could directly link the lib in my package ?

I guess the latter _should_ work too.
Give it a try and let me know :)

**Comment by Kyome22** [comment]
@paescebu I take my hat off to your great achievement! 
Your implemented `SwiftTasksVision` uses XCFramework internally, but how did you build it? 
I am also trying to build it based on precedents (#4517), but it won't build due to compatibility issues with `build_bazel_rules_apple` and `build_bazel_rules_swift`. I would appreciate any hints.

**Comment by paescebu** [comment]
@Kyome22 
Hey!
Thanks for your nice words. Much appreciated.
The good thing about my Swift Package is that it self updates within 24 hours as soon as Google updates the underlying CocoaPod (So im relying on CocoaPods, not bazel builds). This means that there's no secret on how I do it. All the steps are essentially in the separate library target that executes every 24 hours:
https://github.com/paescebu/SwiftTasksVision/blob/main/Sources/UpdatePackage/UpdatePackage.swift
That's literally all the steps involved.

In short: once i figured out how to get it to work manually, it only made sense to automate it :)


And yes, I'm partly reusing and partly repackaging into new XCframeworks (especially the static C libraries) in the end product.
Hope that helps

**Comment by mihaidimoiu** [comment]
@paescebu this is wonderful! Really good job! 
Maybe can we create a github action and create a PR to Google to migrate to SPM? 
In that action build the bazel and then create the package. 

eg: `  FRAMEWORK_NAME=MediaPipeTasksCommon MPP_BUILD_VERSION=0.10.0 \
    ./mediapipe/tasks/ios/build_ios_framework.sh`

**Comment by paescebu** [comment]
@mihaidimoiu Hey!
Unfortunately creating a PR based off of my solution is still not ideal. Cause my solution still relies on some unsafeFlags in the Swift package which comes with some limitations. I think for the solution to qualify for something even "remotely official support" the XCFrameworks would need to be packaged up a bit differently in order to be able to be used "properly" with Swift packages without the currently existing limitations.
TLDR: I honestly think its not good enough (yet) to be used as the official solution.

Obviously we're still all hoping for Google to officially support SPM to simplify integration for all of us. 
But until then you can use my repo, it automatically tracks and updates within 24hours when Google pushes updates to the MediaPipeTasks Pods.

**Comment by mihaidimoiu** [comment]
Hello, @paescebu !

After further research, I've created a [fork](https://github.com/mihaidimoiu/mediapipe) to integrate a package built directly for Swift Package Manager. 

It is not completely perfect, but it is a start to the migration for SPM. Currently, it build from a MacOS with 26.1.

Also [here](https://github.com/mihaidimoiu/mediapipe/releases/tag/v0.10.26) are the artifacts to test for releasing the packages.

Currently, it can be imported correctly in an app, but it doesn't compile the app, since it needs more libs packed, like proto, graphs, and so on

I will continue to investigate but to pack everything together. 

In CocoaPods, there is a flag `OTHER_LDFLAGS` where it can link libs.

**Comment by paescebu** [comment]
@mihaidimoiu Hey!
Quite astonishing the amount of work you put into this already!

Feel free to steal steps from my package update target for your solution. Cause i needed to repackage the static C libraries into a new XCFramework so that i can use them in a Swift package.

I think thats the piece you were missing right? My package update code shows how it was done for my solution :).

But it was surely not enough to take the xcframeworks from the CocoaPods as is, that I remember like it was yesterday 😂

Regards
Pascal

**Comment by Kyome22** [comment]
I've implemented an auxiliary tool that references @paescebu's repository to fetch and build specified versions of XCFrameworks, including past ones. I've checked the linker settings for the minimum required frameworks, but `unsafeFlag` seems to be absolutely necessary.

https://github.com/Kyome22/fetch-mediapipe

**Comment by paescebu** [comment]
Totally, I haven't found a way yet to make it work without unsafeFlags :) 
Feels like the last missing piece to make the solution "truly perfect"

**Comment by mihaidimoiu** [comment]
Hello! 

I have opened a [PR](https://github.com/google-ai-edge/mediapipe/pull/6171). This does not uses "unsafeFlags" anymore. [Here](https://github.com/google-ai-edge/mediapipe/pull/6171/files#diff-953cabcd7b9c4e83bee3b135a2a982e1a421b5a028025b7b0a31249e9c6f5f35) is what I found. I added as dependency those libs. 

Currently, I've managed to import the package and compiled successfully and app that uses Task Vision.

**Comment by DenisovAV** [comment]
I maintain [flutter_gemma](https://github.com/DenisovAV/flutter_gemma), a Flutter plugin for on-device Gemma inference using MediaPipeTasksGenAI. SPM support would allow us to drop the CocoaPods dependency entirely. With CocoaPods going read-only this year, this is becoming a blocker for the iOS ecosystem.

— [Sasha Denisov](https://github.com/DenisovAV)

**Comment by kanaukou-google** [comment]
As of [50f0039](https://github.com/google-ai-edge/mediapipe/commit/50f0039543b1133f1d46842a67b504dbc4adf234), SPM is finally available for MediaPipe tasks. 

Thank you for your patience!

**Comment by mikeg0184** [comment]
Thanks @kanaukou-google. I'm seeing the following issue trying to get a release build out: 

`This bundle Payload/MyApp.app/Frameworks/MediaPipeTasksAudio.framework is invalid. The Info.plist file is missing the required key: CFBundleVersion.`
